Controllers/ArtefactCommentsController.cs

```diff
@@ -19,12 +20,68 @@ public class ArtefactCommentsController : ControllerBase
     {
         private readonly ApplicationDbContext _context;
         private readonly UserService _userService;
+        private readonly IConverter<ArtefactComment> _converter;
 
         public ArtefactCommentsController(ApplicationDbContext context,
-                                UserService userService)
+            UserService userService,
+            IConverter<ArtefactComment> converter)
         {
             _context = context;
             _userService = userService;
+            _converter = converter;
+        }
+
+        // Return all comments of 'artefactId' as a json tree.
+        [HttpGet]
+        public async Task<IActionResult> GetArtefactComments(
+            [FromQuery] string artefactId
+        )
+        {
+            var artefact = await _context
+                .Artefacts
+                .SingleOrDefaultAsync(a => a.Id == artefactId);
+            if (artefact == null)
+            {
+                return NotFound();
+            }
+
+            var artComments = await _context
+                .ArtefactComments
+                .Where(ac => ac.ArtefactId == artefactId)
+                .Include(ac => ac.ChildComments)
+                .ToListAsync();
+
+            var rootComments = artComments
+                .Where(ac => ac.ParentCommentId == null);
+
+            foreach (var comment in rootComments)
+            {
+                AttachChildren(comment);
+            }
+
+            return new JsonResult(rootComments.Select(c => _converter.ToJson(c)));
+
+            // recursively attach children
+            void AttachChildren(ArtefactComment parent) 
+            {
+                List<ArtefactComment> children;
+
+                // check for 0 in case 'parent' is a rootComment
+                if (parent.ChildComments == null || 
+                    parent.ChildComments.Count() == 0) 
+                {
+                    children = artComments
+                        .Where(ac => ac.ParentCommentId == parent.Id)
+                        .ToList();
+
+                    parent.ChildComments = children;
+                }
+
+                foreach (var child in parent.ChildComments) 
+                {
+                    AttachChildren(child);
+                }
```

Comment:
Correct me if I'm wrong, but the idea is that the top `if` is always true right? Unless something funky is happening with the actual data in the database. 
If so I think we should move the lower foreach inside the if, purely for the reason that if there *is* funky database stuff happening, that will prevent infinite recursion.

Reply:
The (not at all well communicated) idea was that if "parent" is an element of "rootComments", then there's no need to attach it's children, since they will have already been attached from the ".Include()".

It certainly would have been helpful it my comment had mentioned that!

Reply:
Ohh right I get it. Personally I dunno I'd just skip the `.Include` but that's immaterial.

I believe my point about the infinite recursion still stands, so I still think the `foreach` should be moved.